cbauthx/cbauth.go

```diff
@@ -0,0 +1,437 @@
+	var failedEndpoints []string
+	var failedErrors []error
+	for {
+		availableEndpoints = availableEndpoints[:0]
+
+		a.lock.Lock()
+		for _, address := range endpoints {
+			if !slices.Contains(failedEndpoints, address) {
+				availableEndpoints = append(availableEndpoints, address)
+			}
+		}
+		a.lock.Unlock()
+
+		if len(availableEndpoints) == 0 {
+			break
+		}
+
+		// Select the primary endpoint if its one of the valid endpoints to use,
+		// otherwise, randomsly select an endpoint to connect to.
+		endpoint := availableEndpoints[rand.Intn(len(availableEndpoints))]
+
+		a.logger.Debug("attempting to build new cbauth client",
+			zap.String("endpoint", endpoint))
+
+		connectCtx, cancel := context.WithTimeout(ctx, a.connectTimeout)
+		defer cancel()
+
+		cli, err := NewCbAuthClient(connectCtx,
+			&CbAuthClientOptions{
+				Logger:        a.logger,
+				HttpTransport: a.httpTransport,
+
+				Endpoint:          endpoint,
+				ServiceName:       a.serviceName,
+				UserAgent:         a.userAgent,
+				Username:          username,
+				Password:          password,
+				ClusterUuid:       clusterUuid,
+				HeartbeatInterval: a.heartbeatInterval,
+				HeartbeatTimeout:  a.heartbeatTimeout,
+				LivenessTimeout:   a.livenessTimeout,
+			})
+		if err != nil {
+			failedEndpoints = append(failedEndpoints, endpoint)
+			failedErrors = append(failedErrors, err)
+
+			a.logger.Warn("failed to build new cbauth client",
+				zap.Error(err))
+			continue
+		}
+
+		return cli, endpoint, nil
+	}
+
+	a.logger.Debug("failed to connect to all cbauth endpoints...")
+	return nil, "", &MultiConnectError{
+		Reasons: failedErrors,
+	}
+}
+
+func (a *CbAuth) Reconfigure(config *CbAuthConfig) error {
+	a.lock.Lock()
+	a.endpoints = config.Endpoints
+	a.username = config.Username
+	a.password = config.Password
+	if config.ClusterUuid != "" {
+		a.clusterUuid = config.ClusterUuid
+	}
+
+	reconfigureSigCh := a.reconfigureSigCh
+	a.reconfigureSigCh = make(chan struct{})
+
+	a.lock.Unlock()
+
+	close(reconfigureSigCh)
+
+	return nil
+}
+
+func (a *CbAuth) runThread(
+	initialCli *CbAuthClient,
+	initialEndpoint string,
+	initialUsername string,
+	initialPassword string,
+	initialClusterUuid string,
+) {
+	var lock sync.Mutex
+
+	cli := initialCli
+	endpoint := initialEndpoint
+	username := initialUsername
+	password := initialPassword
+	clusterUuid := initialClusterUuid
+
+	go func() {
+		for {
+			<-a.reconfigureSigCh
+			a.logger.Debug("cbauth received reconfigure signal")
+
+			lock.Lock()
+			a.lock.Lock()
+
+			// decide whether we need to close the original client
+			if a.closed ||
+				!slices.Contains(a.endpoints, endpoint) ||
+				username != a.username ||
+				password != a.password ||
+				clusterUuid != a.clusterUuid {
+
+				a.logger.Debug("cbauth is invalidating its current client")
+				cli.Close()
+			}
+
+			a.lock.Unlock()
+			lock.Unlock()
+
+			if a.closed {
+				break
+			}
+		}
+	}()
+
+	for {
+		if initialCli != nil {
```

Comment:
Confused what this is doing, surely cli is already assigned to initialCli if initialCli isn't nil?

Reply:
This is true, but I wanted the logic to be more consolidated.  I'll adjust the variable declaration to no longer default to initialCli though.